src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/X509Pal.PublicKey.cs

```diff
@@ -225,7 +227,7 @@ private static byte[] ConstructDSSPublicKeyCspBlob(byte[] encodedKeyValue, byte[
 
             bw.Write(q);
             if (DSS_Q_LEN > cb)
-                bw.Write(new byte[DSS_Q_LEN - cb]);
+                bw.Write(s_emptyBuffer.AsSpan(..(DSS_Q_LEN - cb)));
```

Comment:
Please use the (0, length) instead of the .. version.  I find it much, much more legible.  (Ideally this guidance is everywhere, it's definitely for anything in cryptography)

Reply:
If nothing else, the .. syntax in this specific case has a tiny bit more overhead. From a readability perspective, I also agree in this case it doesn't help. (In some other cases, it's both as cheap and does help with readability.)

Reply:
Okay, I'll change it!
@stephentoub Where can I found out more about the performance implications of using Ranges? As in where it does and where it doesn't add overhead?

Reply:
> As in where it does and where it doesn't add overhead?

It depends on the situation.

The biggest potential cost (which isn't relevant in this specific case) is allocation, where using a range might lead to an allocation in a subwhat hidden fashion, e.g. consider this code:
```C#
ReadOnlySpan<char> ros = input[2..4];
```
Does that allocate?  It depends on the type of `input`.  If `input` is a `ReadOnlySpan<char>`, then no.  But if `input` is a `string`, then this is going to be the equivalent of `input.Substring(2, 2)`, which obviously allocates.  This isn't an issue with ranges per se, it's just something to keep in mind.

The "tiny bit more overhead" I was referring to in this case is that writing:
```C#
array.AsSpan(..x);
```
is compiled into:
```C#
array.AsSpan(Range.EndAt(x));
```
which inlines to become:
```C#
array.AsSpan(new Range(new Index(0), x));
```
which has some code that's hard for the JIT to completely get rid of, as it strives to generate code equivalent to what you can write directly:
```C#
array.AsSpan(0, x);
```
To be clear, it's very minor, but for a case like this where there's arguably no readability/maintainability to the new syntax, we prefer to use the simple code that's clear and doesn't have potential overhead, no matter how small.

Reply:
Makes perfect sense now, thank you for the writeup! Has there been discussion of lowering Ranges in certain scenarios to their (from, to) counterpart within roslyn? I tried searching for an issue in the roslyn repo, but couldn't find anything.

Reply:
> Has there been discussion of lowering Ranges in certain scenarios to their (from, to) counterpart within roslyn?

They are.  You can use the `[x..y]` syntax on a collection type that has a `Slice(int offset, int length)` method, and the compiler will generate the relevant arguments to Slice, e.g.
https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQBgAIUMA6AYQHsAbauAYxgEtKA7KAblwOIwBYuObgGZiaQuUK4A3rkLziopqxgTKAVxWFphAOZwYHQgF85ClKMkBlak3pwAFMtWUAZq6gGANIWeE6rLowABYAlIQAvAB8hKxwAO4SDqGCpkJ4ouiEAGKUlDJm8ha8AGzEfIQAogAeEGAADnTJhdotCoQAbhAIhPQYkbEJSSltCl099OIRvRgA2mSkfAC6gu1pabib6Vu4rHVwUPUQ9rwFOO3FsAjqjIQAkqwAJnDVLbLn7eaZGGVMDbb0JgueqICAwSg9B7PapOLRMcLRQjPVwQdTUQyjIpKLQAcQMAHl3J4YLDVF1qOo4AiYsjUejVgo0u0WpcYNdbgAlCCBOAKN6YxSELk8hxQl6EWDdGA+MXVQhwJ6haRMz5Y+5PcVWGBS7R6AxGFWq4qyqpPXX6QwmFrrIA

Reply:
Ah I vaguely remember reading something about that.
Thank you for always taking the time to elaborate so well! I am very grateful :)